core/templates/services/oppia-rte-parser.service.spec.ts

```diff
@@ -0,0 +1,136 @@
+
+/**
+ * @fileoverview Spec for service that parses rich text string.
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { OppiaRteNode, OppiaRteParserService, TextNode } from './oppia-rte-parser.service';
+
+describe('RTE parser service', () => {
+  let rteParserService: OppiaRteParserService;
+
+  const compareRteNodeToObject = (node: OppiaRteNode, dom) => {
+    const dfs = (n: OppiaRteNode | TextNode, d) => {
+      if ('value' in n) {
+        if (!d.value) {
+          return false;
+        }
+        if (n.value !== d.value) {
+          return false;
+        }
+        return true;
+      }
+      if (d.value) {
+        return false;
+      }
+      if (n.selector !== d.tag) {
+        return false;
+      }
+      if ((n.attrs && !d.attrs) || (!n.attrs && d.attrs)) {
+        return false;
+      }
+      if (n.attrs && d.attrs) {
+        if (Object.keys(n.attrs).length !== Object.keys(d.attrs).length) {
+          return false;
+        }
+        for (let key of Object.keys(n.attrs)) {
+          if (n.attrs[key] !== d.attrs[key]) {
+            return false;
+          }
+        }
+        if (n.selector.startsWith(rteParserService.NON_INTERACTIVE_PREFIX)) {
+          return true;
+        }
+      }
+
+      if (n.children.length !== d.children.length) {
+        return false;
+      }
+      for (let childIndex = 0; childIndex < n.children.length; childIndex++) {
+        if (!dfs(n.children[childIndex], d.children[childIndex])) {
+          return false;
+        }
+      }
+      return true;
+    };
+    return dfs(node, dom);
+  };
+
+  beforeEach(() => {
+    rteParserService = TestBed.inject(OppiaRteParserService);
+  });
+
+  it('should generate a custom representation of the rte string', () => {
+    let testCases = [{
+      rteString: '<p>Hi<em>Hello</em>Hello</p>' +
+                 '<oppia-noninteractive-link url-with-value="url"' +
+                                            'link-with-value="link">' +
+                 '</oppia-noninteractive-link>',
+      representation: {
+        tag: 'body',
+        attrs: {},
+        children: [
+          {
+            tag: 'p',
+            attrs: {},
+            children: [
+              new TextNode('Hi'),
+              {
+                tag: 'em',
+                attrs: {},
+                children: [new TextNode('Hello')]
+              },
+              new TextNode('Hello')
+            ]
+          },
+          {
+            tag: 'oppia-noninteractive-link',
+            attrs: {
+              urlWithValue: 'url',
+              linkWithValue: 'link'
+            }
+          }
+        ]
+      }
+    }];
+    testCases.forEach(testCase => {
+      let node = rteParserService.constructFromRteString(testCase.rteString);
+      expect(
+        compareRteNodeToObject(node, testCase.representation)
+      ).toBe(true, testCase.rteString);
+    });
+  });
+
+  // Noninteractive components are easily identified by the oppia-noninteractive
```

Comment:
What could possibly happen if they use a valid tag but invalid params? Can that cause vulnerabilities?

Reply:
That part is handled in the rte-output component. Since we are manually typing out all the attributes, random attributes will be ignored. Hence no vuls here.
<img width="1024" alt="Screenshot 2021-11-26 at 10 36 49" src="https://user-images.githubusercontent.com/23002114/143567671-9f1c22d4-f66d-4a44-9f9c-1b9993d36163.png">